### ISSUE TITLE
Motors bending backwards when running pre-trained policy.

### ISSUE THREAD

**Issue by Dharmik-P45** [leap-hand/LEAP_Hand_Sim#19]
After running `roslaunch example.launch`, the hand does go to the home pose. But after running the "rotate cube" pre-trained policy, the hand goes to some initial state that seems 60% correct, but not exactly like holding a cube. When the model starts running the 3 fingers move backwards until they're unable to, and the thumb makes small micro-movements.

![initial position](https://github.com/leap-hand/LEAP_Hand_Sim/assets/113470137/1052c22f-813f-4d25-a9d3-b025e9e33da2)
![after some time](https://github.com/leap-hand/LEAP_Hand_Sim/assets/113470137/85877ae8-ed5b-4b67-a0a9-fec724729249)

I've used the leap-hand API to manually control the hand and it works perfectly, so I don't suspect any ROS/dynamixel//assembly errors.

One thing I noticed is that sampling one of the commands during the run, and plugging that into the manual leap-hand controller produces a different pose than what the hand is doing when the models running. The pose produced is still an awkward pose though and not close to a grasping.

Running the grasp pre-trained policy bring it to a similar initial state, then does no additional movement at all, while the script continues to produce commands.

**Comment by Lucky7Chess** [comment]
Hmm it is difficult for me to know what the issue is remotely.  Are you sure the thumb is not rotated 90/180 degrees?    You also need a cube in the hand, otherwise if it doesn't sense a cube it will definitely not work.

**Comment by Dharmik-P45** [comment]
> Hmm it is difficult for me to know what the issue is remotely. Are you sure the thumb is not rotated 90/180 degrees? You also need a cube in the hand, otherwise if it doesn't sense a cube it will definitely not work.

I'm pretty sure the thumb isn't rotated by any amount. Because the home position works perfectly, as well as manual control. I just checked each joint by setting them to 45 deg and they all moved as expected.

I don't have a 7.5cm cube but I did try using similar sized objects - the pre-grasp pose didn't even come close to touching the objects, the first photo is the pre-grasp and shows the two fingers and thumb aren't doing anything meaningful. If I were to put the real size cube in there it would just fall down.

The issacgym simulation runs perfectly if that helps.

EDIT: If there was a video or something of running the model from the start it would be helpful to compare, especially what the pre-grasp pose is supposed to look like.

**Comment by Lucky7Chess** [comment]
The pregrasp being wrong makes me think there's something off with the home pose, but if you're sure there's not then I'm.not sure what else the issue is.  For us it basically starts at the grasp pose and then rotates as the video we have in the website.  If we remove the cube the fingers will open up and it will fail.

**Comment by Dharmik-P45** [comment]
> The pregrasp being wrong makes me think there's something off with the home pose, but if you're sure there's not then I'm.not sure what else the issue is. For us it basically starts at the grasp pose and then rotates as the video we have in the website. If we remove the cube the fingers will open up and it will fail.


Is the command to set the pose during the policy the same format/order as the set_allegro([]) command when doing it manually using leaphand SDK? Like I said the model outputs a command of joint angles and the hand moves to an awkward position. When I plug those same angles into a script to manually control the hand, a different pose happens, (still awkward) even though its the same exact command.

**Comment by Lucky7Chess** [comment]
There's this reordering that happens between Isaac gym and the real hand because Isaac defines random order.  Are you running exactly the example code?  You may need to dig into the code mode here to find the big.

**Comment by Dharmik-P45** [comment]
> There's this reordering that happens between Isaac gym and the real hand because Isaac defines random order. Are you running exactly the example code? You may need to dig into the code mode here to find the big.

Yeah I'm running the ros example.launch first, then the provided command for deploy.py, without changing any code. Should I be looking for bugs related to where the commands are coming from like looking into the model itself, or look at how those commands are executed in the hand?

Thanks for the help!

**Comment by Lucky7Chess** [comment]
If you're running our code, with the same Isaac gym version, then it should work exactly.  I would try going joint by joint and making sure the joint order is the same as on our computer.  Isaac has that weird bug with the joint order where the urdf and the way it loads does not match.

**Comment by Dharmik-P45** [comment]
> I would try going joint by joint and making sure the joint order is the same as on our computer.

How do I see what the order is on your computer? I found the URDF file and the code snippet that iterates through all the joints. Printing each joint that's processed just prints 0->15 in order. This is probably not what you're talking about though so I'd appreciate some clarification on that.

I do see that joint_0 in the URDF is palm_lower_left, while dynamixel motor 0 controls the index finger's first joint. I'm not sure if those are the same joints.

**Comment by Lucky7Chess** [comment]
I think if you go to the Isaac gym gui/viewer it'll number the joints and they won't be 0-16.

**Comment by Dharmik-P45** [comment]
> I think if you go to the Isaac gym gui/viewer it'll number the joints and they won't be 0-16.

There are 17 joints/parts labeled 0-16. The extra one is palm_lower which doesn't act as a joint/doesn't change orientation/position. So the moveable joints are labeled 1-16, while the URDF file and everything else labels them 0-15. Not sure if this is an issue.

I'll try with a 7..5 cm cube and see what ahppens.

![gym](https://github.com/leap-hand/LEAP_Hand_Sim/assets/113470137/c02a7057-cc6f-44f6-afc8-c498d8ca3a32)

**Comment by Dharmik-P45** [comment]
Side question: How would you recommend going about training an in-hand rotation for anything besides cubes, especially an irregular shaped object?

Do the current tools available in leap_hand_sim allow for generating those types of grasps? How?

**Comment by Lucky7Chess** [comment]
The screenshot you have is bodies not joints.

You can upload more urdf/models of different objects yes and train on those as well.  Same pipeline should work but that's beyond the scope of this repo.

**Comment by zahid990170** [comment]
@Lucky7Chess @Dharmik-P45 

Having the similar problem when running the deploy script. The home pose works ok ,i.e., when running

`roslaunch example.launch`, the hand correctly takes the home pose. 

I wanted to confirm if there can be some issue regarding the indices of joints between sim and real. 

`LeapHandRot.yaml` lists 
`sim_to_real_indices: [1, 0, 2, 3, 9, 8, 10, 11, 13, 12, 14, 15, 4, 5, 6, 7]`

I have the leap hand (right hand), and the motors have been given IDs accordingly to this urdf 

[onshape.pdf](https://github.com/user-attachments/files/15536238/onshape.pdf)

Thumb: joint index 12, 13, 14, 15
Finger 3 (finger opposite to the thumb): joint index 8, 9, 10, 11
Finger 2 (next to finger 1): joint index 4, 5, 6, 7
Finger 1 (next to the thumb): joint index 0, 1, 2, 3

**Comment by Lucky7Chess** [comment]
I'm talking about the IDs that map between Isaac and the Leap hand itself.  Check that.  Make sure that you can mirror the same pose in sim and real.

**Comment by Lucky7Chess** [comment]
I also wonder if there's some hardware variation in the motors depending on the batch you get.  May have to randomize a bit more.

**Comment by zahid990170** [comment]
@Lucky7Chess 

![Screenshot from 2024-06-03 17-35-30](https://github.com/leap-hand/LEAP_Hand_Sim/assets/56445034/4a8451ab-77a7-4d27-8281-789b2b89a6cd)

are these the IDs for Isaac ? But here these are completely different from the urdf visualization.

**Comment by Lucky7Chess** [comment]
Dofs, not bodies tab.

**Comment by Lucky7Chess** [comment]
Are you using the newest Isaac gym preview?

**Comment by Lucky7Chess** [comment]
Make sure the cube you are using is exactly the same size and weight as in simulation also.

**Comment by zahid990170** [comment]
> Dofs, not bodies tab.

![image](https://github.com/leap-hand/LEAP_Hand_Sim/assets/56445034/d27b6d7b-01a6-43f3-b290-266277294946)

**Comment by zahid990170** [comment]
> Make sure the cube you are using is exactly the same size and weight as in simulation also.

The dimension of the side is 7.5 cm. Its 3D printed. What should be the weight?

**Comment by zahid990170** [comment]
> Are you using the newest Isaac gym preview?

This is the version I am using: 
![image](https://github.com/leap-hand/LEAP_Hand_Sim/assets/56445034/b6bed774-69bb-4b25-a192-9dcaf69666c4)

**Comment by Lucky7Chess** [comment]
Have to do dof pipeline on CPU.  The weight has to match what you set in urdf.

**Comment by zahid990170** [comment]
> Have to do dof pipeline on CPU. The weight has to match what you set in urdf.

I checked and motors IDs (joint indices) match between isaacgym and leaphand. 

In cube.urdf, the mass was set to 50 g. 
` <mass value="0.05" /> `

The cube was printed with the same specs.

**Comment by zahid990170** [comment]
Below are some images, 
first one is the home position that leap hand takes after running, `example.launch`

![1717507339884](https://github.com/leap-hand/LEAP_Hand_Sim/assets/56445034/2731726f-577c-4da1-974a-fffcbc6a8e80)


and second one is the first pose that the hand takes after deploy script is launched. 
![1717507357379](https://github.com/leap-hand/LEAP_Hand_Sim/assets/56445034/39308549-2120-481c-8768-4a7387533d80)

home pose and the cube in hand
![1717507598753](https://github.com/leap-hand/LEAP_Hand_Sim/assets/56445034/996b5264-1c6a-481f-bdda-d3044e7dac1d)

and a few moments later, the fingers go backwards. I stopped the deploy script. 
![1717507766385](https://github.com/leap-hand/LEAP_Hand_Sim/assets/56445034/9a4369db-c280-4eea-9990-ed0c30720e38)

**Comment by Dharmik-P45** [comment]
> I'm talking about the IDs that map between Isaac and the Leap hand itself. Check that. Make sure that you can mirror the same pose in sim and real.

There are joints labeled from 0-15, but they seem out of order. For example joint 0 on Isaac is the abduction/adduction of the index finger, while on the leap hand joint 0 is the bottom-most joint for index finger. It seems like the ab/adduction joint and the bottom joints are swapped for all fingers besides the thumb. The URDF matches Isaac so the discrepancy is between the URDF/IsaacSim and the actual leap hand motor IDs. 

I'll search the code for a fix but what do you think needs to be changed to fix this? The URDF file?

**Comment by Lucky7Chess** [comment]
Are the motors mirrored in their mounting.  They seem to be opening a lot more than I'd expect them to.

**Comment by Dharmik-P45** [comment]
No I don't think they're mirrored. Do you remember what part of the code contains the condition where it opens up the hand if an object isn't detected? I want to try removing that condition so the hand rotates something even if the object isn't physically there.

**Comment by dongwflj** [comment]
I got the same problem on real palm, can I ask what is motor id order map to the palm. I set the motor id as the urdf but still not work:
thumb:12-13-14-15
next thumb finger:1-0-2-3
middle finger:5-4-6-7
last finger:9-8-10-11

**Comment by Lucky7Chess** [comment]
You can't do that @Dharmik-P45 , that's part of the RL policy.

**Comment by Lucky7Chess** [comment]
You may need to tune this slightly based on your hand.  Make sure the motors are mounted and work correctly and you're using XC motors.

**Comment by dongwflj** [comment]
I add below code in deploy.py and seems it works with right finger pose, but it can't rotate cube. the reason I change the code is the init pose(prev_target varibe) is a real value (not in [-1, 1]), but the action is a sim value from Isaacgym, so directly add the 2 value is wired, I just map the action's sim value to real value and then add with prev_target, not sure it is correct:
-----------------------------------------------------------------        
         def scale(x, lower, upper):
            return (0.5 * (x + 1.0) * (upper - lower) + lower) 

          action = scale(action, self.leap_dof_lower, self.leap_dof_upper) #Add this line
          print("Action after scale", action)
          if "actions_mask" in self.config["task"]["env"]:
              action = action * torch.tensor(self.config["task"]["env"]["actions_mask"]).cuda()[None, :]
          print("prev target:", prev_target)
          target = prev_target + self.action_scale * action 
          target = torch.clip(target, self.leap_dof_lower, self.leap_dof_upper)
          prev_target = target.clone()

**Comment by soham2560** [comment]
@Dharmik-P45 were you able to resolve this issue?

**Comment by lixiaolong04** [comment]
> 您能够解决此问题吗？

I am also trying to solve this problem. My launch file works well, but my deploy file does not. Can we talk about it together? Thank you

**Comment by Dharmik-P45** [comment]
> @Dharmik-P45 were you able to resolve this issue?

@soham2560  I was not. My work with the LEAP hand has since ended but I can help out if needed.
